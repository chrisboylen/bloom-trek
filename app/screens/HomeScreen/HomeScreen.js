import React, { Component } from 'react';
import { View, StyleSheet } from 'react-native';
import Map from '../../components/Map/Map';
import Search from '../../components/Search/Search';
import { Location, Permissions } from 'expo';
import { getAllParks } from '../../utils/api';
import Geocoder from 'react-native-geocoding';
import { GOOGLE_KEY } from '../../../key';

Geocoder.init(GOOGLE_KEY);

export default class HomeScreen extends Component {
  state = {
    location: null,
    errorMessage: null,
    parks: [],
    selectedPark: null,
  };

  componentDidMount() {
    this.getCurrentLocation();
    this.getParks();
  }

  getCurrentLocation = async () => {
    const { status } = await Permissions.askAsync(Permissions.LOCATION);
    if (status !== 'granted') {
      this.setState({
        errorMessage: 'Permission to access location was denied...',
      });
    }

    const position = await Location.getCurrentPositionAsync({});
    const location = {
      latitude: position.coords.latitude,
      longitude: position.coords.longitude,
    };
    this.setState({ location });
  };

  getParks = async () => {
    const unformattedParks = await getAllParks();
    const parks = unformattedParks.map(park => ({
      id: park.id,
      name: park.name,
      coords: {
        latitude: park.latitude,
        longitude: park.longitude,
      },
    }));
    this.setState({ parks });
  };

  handleSelectedPark = (id) => {
    this.setState({ selectedPark: id });
	};
	
	updateLocation = (input) => {
    Geocoder.from(input)
      .then((json) => {
        const { location } = json.results[0].geometry;
        const { lat, lng } = location;
        this.setState({
          location: {
						latitude: lat,
						longitude: lng
					}
        });
      })
      .catch(error => console.warn(error));
  };

  render() {
    const { location, parks } = this.state;

    return (
      <View style={styles.container}>
				<Search updateLocation={this.updateLocation} />
        <Map
          location={location}
          parks={parks}
          handleSelectedPark={this.handleSelectedPark}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 7,
  },
});
