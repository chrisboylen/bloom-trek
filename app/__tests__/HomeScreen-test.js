import React from 'react';
import { shallow } from 'enzyme';
import HomeScreen from '../screens/HomeScreen';

/* global spyOn */

describe('HomeScreen', () => {
  let unformattedParks;
  let wrapper;

  beforeEach(() => {
    unformattedParks = [{
      id: 1,
      name: 'Castlewood Canyon',
      coords: {
        latitude: 39.3379,
        longitude: -104.7512,
      },
    }];
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve(unformattedParks),
    }));
    wrapper = shallow(<HomeScreen />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('getParks', () => {
    it('should set state with parks', async () => {
      await wrapper.instance().getParks();

      expect(wrapper.state().parks).toEqual(unformattedParks);
    });
  });

  describe('handleSelectedPark', () => {
    it('should set state when invoked ', async () => {
      wrapper.setState({ parks: unformattedParks });
      await wrapper.instance().handleSelectedPark(1);

      expect(wrapper.state().selectedPark).toEqual(1);
      expect(wrapper.state().selectedTrail).toEqual(null);
      expect(wrapper.state().trails).toEqual(unformattedParks);
    });
  });

  describe('resetMap', () => {
    it('should invoke getCurrentLocation when called', () => {
      const spy = spyOn(wrapper.instance(), 'getCurrentLocation');

      wrapper.instance().resetMap();

      expect(spy).toHaveBeenCalled();
    });
  });
});
