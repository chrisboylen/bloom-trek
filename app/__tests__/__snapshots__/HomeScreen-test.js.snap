// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HomeScreen />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "accessible": true,
      "allowFontScaling": true,
      "children": "broken",
      "ellipsizeMode": "tail",
    },
    "ref": null,
    "rendered": "broken",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "accessible": true,
        "allowFontScaling": true,
        "children": "broken",
        "ellipsizeMode": "tail",
      },
      "ref": null,
      "rendered": "broken",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
